### ISSUE TITLE
Switch from versioneer to setuptools_scm

### ISSUE THREAD

**Issue by kmpaul** [kmpaul/mpirical#16]
MPIrical currently uses [versioneer](https://github.com/warner/python-versioneer) for versioning, but many newer packages use [bumpversion](https://github.com/peritus/bumpversion).  Unfortunately, both have not been updated in a long time, and the current bumpversion GitHub README actually says that you should use [bump2version](https://github.com/c4urself/bump2version) instead, because it is still maintained.

| PyPI Package | Last Release   | Stars  | Used by | Tests     |
|---------------|---------------|-------|----------|--------|
| versioneer       | Jan 1, 2017     | 645    | 305         | Passing |
| bumpversion   | Apr 15, 2015   | 1279  | 8783       | Failing   |
| bump2version | Aug 26, 2019 | 268    | 826         | Passing |

Based on these statistics, it seems to me that bumpversion is clearly the more popular paradigm, and that bump2version is the correct fork to use for current work.

Should we switch?

**Comment by kmpaul** [comment]
@andersy005 Do you have any comments to make on this issue?  It seems that many people feel that versioneer is "too invasive."

**Comment by andersy005** [comment]
For intake-esm, and other packages we used to have `versioneer`. We've switched to using [setuptools_scm](https://github.com/pypa/setuptools_scm/) which integrates very well with `setuptools`. I personally find it very easy to work with, and the fact that it is under the Python Packaging Authority group provides some "guarantee" in regards to its maintainability.

**Comment by kmpaul** [comment]
Yeah.  Okay.  I'm willing to use/move to `setuptools_scm`.  However, the Python Packaging Authority isn't really a selling point.  They are also the people who claim that `pipenv`+`virtualenv` should be used instead of `conda`.  I am a bit concerned that `setuptools` and `pip` will evolve to suit the needs to `pipenv` and `virtualenv`, in competition with `conda`.  Do you have any concerns about this?

**Comment by kmpaul** [comment]
Actually, I guess the PyPA [recognizes conda](https://packaging.python.org/key_projects/#conda) as a valid solution to packaging.  I think I'm just having trouble keeping all of the different packages for packaging straight and understanding where conda fits in.  

For example, should we assume that `pip` + `virtualenv` + `wheel` is replaced by `conda`?  Or, should our packages be fully PyPA "compliant" (whatever that means) with `conda` as an additional layer/use-case?

**Comment by andersy005** [comment]
> Yeah. Okay. I'm willing to use/move to `setuptools_scm`. However, the Python Packaging Authority isn't really a selling point. They are also the people who claim that `pipenv`+`virtualenv` should be used instead of `conda`. I am a bit concerned that `setuptools` and `pip` will evolve to suit the needs to `pipenv` and `virtualenv`, in competition with `conda`. Do you have any concerns about this?

Regarding the `pip` vs `conda` debate, I think that Jake Vanderplas' post: [Conda: Myths and Misconceptions](https://jakevdp.github.io/blog/2016/08/25/conda-myths-and-misconceptions/) provides a great, accurate summary. 

> For example, should we assume that pip + virtualenv + wheel is replaced by conda? Or, should our packages be fully PyPA "compliant" (whatever that means) with conda as an additional layer/use-case?

As of [conda v4.6+](https://docs.conda.io/projects/conda/en/latest/user-guide/configuration/pip-interoperability.html), conda provides preview support for better interoperability with pip. 

Per the release notes:

> Conda and pip have historically had difficulties getting along. Pip hasn’t respected Conda’s environment constraints, while Conda has been all too happy to clobber pip-installed software. It’s a mess. Conda 4.6.0 adds preview support for better interoperability. With this interoperability, Conda can use pip-installed packages to satisfy dependencies, and can even remove pip-installed software cleanly and replace them with Conda packages when appropriate. There’s still room for improvement before pip and Conda are hunky-dory BFFs, but we hope this is a good start.

I've been using this preview feature for a while, and it appears to work well for most packages including the ones with wheels. 

> For example, should we assume that pip + virtualenv + wheel is replaced by conda?

As Jake points out in his blog post, you can use conda within your virtualenv today:

```
$ virtualenv test_conda

$ source test_conda/bin/activate

$ pip install conda

$ conda install numpy
```

**Comment by kmpaul** [comment]
Yes.  I've read Jake Vanderplas's post, actually.  But I'm trying to clarify what part of the PyPA-supported stack we should explicitly use (or *not* use) in our software.

For example, in many of our packages, we use *both* `conda` and `pip`.  And this has been made easier through `conda`'s `pip`-interoperability efforts.  However, does that mean we should have *both* a `requirements.txt` file *and* an `environment.yml` file?  Or just a `requirements.txt` file?  And what about `requirements-dev.txt` files?

Should we have a `Pipfile` file to support `pipenv`?  Or do we *only* support `conda`?

There are probably other things, but I've never really grasped the clear delineation between PyPA-stack and the conda-specific stack in our software.  What is the rule?  Start with PyPA-stack (e.g., `setuptools`, `pip`, etc.), and then replace only the parts that `conda` makes redundant (e.g., `pipenv`?  Not sure.

**Comment by jukent** [comment]
@kmpaul Does this relate to our conversation about removing support for Python 2?

**Comment by kmpaul** [comment]
I don't think so.  I think these tools will work with Python 2 as well as Python 3, so it doesn't have to do with the removal of Python 2 support.  I think this is separate.

**Comment by jukent** [comment]
Closed via #19
